### PR TITLE
[Feature] Support the string fucntion TRANSLATE

### LINKED ISSUE #22704: [function]translate
In StarRocks, there is a string function named replace, it can replace a single character with another expression. Still, we need a translate function to do it more flexibly.

here is a proposal:

## description
The function TRANSLATE() is used to substitute characters within a string. It works by taking a string as input, along with a set of characters to be replaced, and the corresponding characters to replace them with. TRANSLATE() then performs the specified substitutions.
just like oracle and snowflake did:
https://docs.oracle.com/database/121/SQLRF/functions232.htm#SQLRF06145
https://docs.snowflake.com/en/sql-reference/functions/translate

## syntax

```
TRANSLATE( <expr>, <from_string>, <to_string> )
```

in which
1. expr:  the expression to be translated. When a character in the expr is not found in the from_string, it is simply included in the result string.
2. from_string: Each character in the from_string is either replaced by its corresponding character in the to_string, or if there is no corresponding character (i.e. if the to_string has fewer characters than the from_string, the character is excluded from the resulting string.)
3. to_string: A string with all characters that are used to replace characters from the from_string.

If to_string is longer than from_string, need to report an error `The third parameter is too long and would be truncated.`

## examples

```
select translate('s1m1a1rrfcks','mf1','to') as test;

+----------------+
| test |
+----------------+
| starrocks           |
+----------------+

```

### PR DESCRIPTION
**Pull request by ZiheLiu**
Closes #22704.

Inspired by and will close #23611.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

File: be/src/exprs/string_functions.cpp [hunk 1]
```diff
@@ -1049,6 +1050,384 @@ StatusOr<ColumnPtr> StringFunctions::repeat(FunctionContext* context, const Colu
...
+
+/**
+ * Translate a string by the ASCII map.
+ *
+ * @param src the string to translate.
+ * @param ascii_map the ASCII map.
+ * @param dst the destination to store the translated chars. The caller must guarantee there is enough room.
+ * @return the number of translated chars.
+ */
+static inline size_t translate_string_with_ascii_map(const Slice& src, const TranslateState::ASCII_MAP& ascii_map,
```
Comment (satanson) [inline be/src/exprs/string_functions.cpp:1154]: it seems that we can use byte array instead of ASCII_MAP.

Comment (ZiheLiu) [reply be/src/exprs/string_functions.cpp:1154]: Great!

File: be/src/exprs/string_functions.cpp [hunk 2]
```diff
@@ -1049,6 +1050,400 @@ StatusOr<ColumnPtr> StringFunctions::repeat(FunctionContext* context, const Colu
...
+    auto& dst_offsets = dst->get_offset();
+    auto& dst_bytes = dst->get_bytes();
+    const auto& src_offsets = src->get_offset();
+
+    const size_t num_rows = src->size();
+    if (num_rows == 0) {
+        return dst;
+    }
+
+    const int num_src_bytes = src_offsets[src_offsets.size() - 1];
```
Comment (fzhedu) [inline be/src/exprs/string_functions.cpp:1248]: `const int num_src_bytes = src_offsets.back();`

**Review by fzhedu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 215 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/string_functions.cpp|0|196|00.00%|[1075, 1077, 1089, 1090, 1091, 1093, 1094, 1096, 1097, 1098, 1099, 1101, 1102, 1103, 1104, 1105, 1106, 1109, 1110, 1111, 1112, 1116, 1118, 1119, 1120, 1121, 1123, 1124, 1139, 1142, 1143, 1144, 1145, 1146, 1147, 1148, 1149, 1151, 1154, 1155, 1157, 1158, 1161, 1172, 1174, 1175, 1176, 1177, 1178, 1179, 1180, 1184, 1187, 1188, 1189, 1192, 1193, 1196, 1197, 1198, 1200, 1202, 1203, 1205, 1206, 1208, 1210, 1213, 1214, 1215, 1216, 1218, 1233, 1235, 1236, 1238, 1239, 1240, 1241, 1243, 1244, 1245, 1248, 1249, 1250, 1251, 1253, 1254, 1256, 1257, 1258, 1259, 1260, 1261, 1264, 1266, 1282, 1284, 1285, 1287, 1288, 1289, 1290, 1292, 1293, 1294, 1297, 1298, 1302, 1303, 1304, 1305, 1307, 1308, 1309, 1311, 1312, 1313, 1315, 1316, 1317, 1319, 1320, 1322, 1323, 1324, 1325, 1326, 1328, 1329, 1333, 1334, 1336, 1349, 1350, 1352, 1353, 1354, 1356, 1357, 1358, 1359, 1361, 1362, 1363, 1366, 1367, 1368, 1369, 1370, 1371, 1373, 1374, 1375, 1376, 1378, 1379, 1380, 1381, 1382, 1383, 1386, 1387, 1388, 1390, 1391, 1392, 1394, 1395, 1396, 1398, 1399, 1400, 1402, 1403, 1405, 1406, 1407, 1408, 1409, 1410, 1412, 1413, 1418, 1419, 1421, 1424, 1425, 1427, 1428, 1429, 1430, 1431, 1433, 1436, 1439]|
|:large_blue_circle:|src/util/utf8_encoding.h|0|19|00.00%|[35, 36, 37, 38, 40, 41, 47, 49, 50, 51, 52, 55, 58, 63, 66, 67, 68, 69, 71]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 215 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/string_functions.cpp|0|196|00.00%|[1075, 1077, 1089, 1090, 1091, 1093, 1094, 1096, 1097, 1098, 1099, 1101, 1102, 1103, 1104, 1105, 1106, 1109, 1110, 1111, 1112, 1116, 1118, 1119, 1120, 1121, 1123, 1124, 1139, 1142, 1143, 1144, 1145, 1146, 1147, 1148, 1149, 1151, 1154, 1155, 1157, 1158, 1161, 1172, 1174, 1175, 1176, 1177, 1178, 1179, 1180, 1184, 1187, 1188, 1189, 1192, 1193, 1196, 1197, 1198, 1200, 1202, 1203, 1205, 1206, 1208, 1210, 1213, 1214, 1215, 1216, 1218, 1233, 1235, 1236, 1238, 1239, 1240, 1241, 1243, 1244, 1245, 1248, 1249, 1250, 1251, 1253, 1254, 1256, 1257, 1258, 1259, 1260, 1261, 1264, 1266, 1282, 1284, 1285, 1287, 1288, 1289, 1290, 1292, 1293, 1294, 1297, 1298, 1302, 1303, 1304, 1305, 1307, 1308, 1309, 1311, 1312, 1313, 1315, 1316, 1317, 1319, 1320, 1322, 1323, 1324, 1325, 1326, 1328, 1329, 1333, 1334, 1336, 1349, 1350, 1352, 1353, 1354, 1356, 1357, 1358, 1359, 1361, 1362, 1363, 1366, 1367, 1368, 1369, 1370, 1371, 1373, 1374, 1375, 1376, 1378, 1379, 1380, 1381, 1382, 1383, 1386, 1387, 1388, 1390, 1391, 1392, 1394, 1395, 1396, 1398, 1399, 1400, 1402, 1403, 1405, 1406, 1407, 1408, 1409, 1410, 1412, 1413, 1418, 1419, 1421, 1424, 1425, 1427, 1428, 1429, 1430, 1431, 1433, 1436, 1439]|
|:large_blue_circle:|src/util/utf8_encoding.h|0|19|00.00%|[35, 36, 37, 38, 40, 41, 47, 49, 50, 51, 52, 55, 58, 63, 66, 67, 68, 69, 71]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 215 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exprs/string_functions.cpp|0|196|00.00%|[1075, 1077, 1089, 1090, 1091, 1093, 1094, 1096, 1097, 1098, 1099, 1101, 1102, 1103, 1104, 1105, 1106, 1109, 1110, 1111, 1112, 1116, 1118, 1119, 1120, 1121, 1123, 1124, 1139, 1142, 1143, 1144, 1145, 1146, 1147, 1148, 1149, 1151, 1154, 1155, 1157, 1158, 1161, 1172, 1174, 1175, 1176, 1177, 1178, 1179, 1180, 1184, 1187, 1188, 1189, 1192, 1193, 1196, 1197, 1198, 1200, 1202, 1203, 1205, 1206, 1208, 1210, 1213, 1214, 1215, 1216, 1218, 1233, 1235, 1236, 1238, 1239, 1240, 1241, 1243, 1244, 1245, 1248, 1249, 1250, 1251, 1253, 1254, 1256, 1257, 1258, 1259, 1260, 1261, 1264, 1266, 1282, 1284, 1285, 1287, 1288, 1289, 1290, 1292, 1293, 1294, 1297, 1298, 1302, 1303, 1304, 1305, 1307, 1308, 1309, 1311, 1312, 1313, 1315, 1316, 1317, 1319, 1320, 1322, 1323, 1324, 1325, 1326, 1328, 1329, 1333, 1334, 1336, 1349, 1350, 1352, 1353, 1354, 1356, 1357, 1358, 1359, 1361, 1362, 1363, 1366, 1367, 1368, 1369, 1370, 1371, 1373, 1374, 1375, 1376, 1378, 1379, 1380, 1381, 1382, 1383, 1386, 1387, 1388, 1390, 1391, 1392, 1394, 1395, 1396, 1398, 1399, 1400, 1402, 1403, 1405, 1406, 1407, 1408, 1409, 1410, 1412, 1413, 1418, 1419, 1421, 1424, 1425, 1427, 1428, 1429, 1430, 1431, 1433, 1436, 1439]|
|:large_blue_circle:|src/util/utf8_encoding.h|0|19|00.00%|[35, 36, 37, 38, 40, 41, 47, 49, 50, 51, 52, 55, 58, 63, 66, 67, 68, 69, 71]|

**Review by satanson — APPROVED** [review]

**Review by fzhedu — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by ZiheLiu** [comment]
@Mergify backport branch-3.0 branch-3.1

**Pull request merged**